stan/math/rev/fun/eigenvalues_sym.hpp

```diff
@@ -0,0 +1,117 @@
+#include <stan/math/rev/fun/value_of_rec.hpp>
+#include <stan/math/rev/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err/check_symmetric.hpp>
+#include <stan/math/prim/err/check_nonzero_size.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+class eigenvalues_vari : public vari {
+ public:
+  int M_;  // A.rows() = A.cols()
+  double *A_;
+  double *w_;  // eigenvalues
+  double *v_;  // eigenvectors
+  vari **vari_ref_A_;
+  vari **vari_ref_w_;
+  vari **vari_ref_v_;
+
+  explicit eigenvalues_vari(const Eigen::Matrix<var, -1, -1> &A)
+      : vari(0.0),
+        M_(A.rows()),
+        A_(reinterpret_cast<double *>(
+            ChainableStack::instance_->memalloc_.alloc(sizeof(double) * A.rows()
+                                                       * A.cols()))),
+        w_(reinterpret_cast<double *>(
+            ChainableStack::instance_->memalloc_.alloc(sizeof(double)
+                                                       * A.rows()))),
+        v_(reinterpret_cast<double *>(
+            ChainableStack::instance_->memalloc_.alloc(sizeof(double) * A.rows()
+                                                       * A.cols()))),
+        vari_ref_A_(reinterpret_cast<vari **>(
+            ChainableStack::instance_->memalloc_.alloc(sizeof(vari *) * A.rows()
+                                                       * A.cols()))),
+        vari_ref_w_(reinterpret_cast<vari **>(
+            ChainableStack::instance_->memalloc_.alloc(sizeof(vari *)
+                                                       * A.rows()))),
+        vari_ref_v_(reinterpret_cast<vari **>(
+            ChainableStack::instance_->memalloc_.alloc(sizeof(vari *) * A.rows()
+                                                       * A.cols()))) {
+    using Eigen::Map;
+
+    Map<matrix_d> Ad(A_, M_, M_);
+    Map<matrix_d> wd(w_, M_, 1);
+    Map<matrix_d> vd(v_, M_, M_);
+    Ad = A.val();
+    Eigen::SelfAdjointEigenSolver<matrix_d> solver(Ad);
+    wd = solver.eigenvalues();
+    vd = solver.eigenvectors();
+
+    Map<matrix_vi>(vari_ref_A_, M_, M_) = A.vi();
+    Map<vector_vi>(vari_ref_w_, M_)
+        = wd.unaryExpr([](double x) { return new vari(x, false); });
+    Map<matrix_vi>(vari_ref_v_, M_, M_)
+        = vd.unaryExpr([](double x) { return new vari(x, false); });
+  }
+
+  /**
+   * Reverse mode differentiation algorithm reference:
+   *
+   * Mike Giles. An extended collection of matrix derivative results for
+   * forward and reverse mode AD.  Jan. 2008.
+   *
+   * Section 3.1 Eigenvalues and eigenvectors.
+   */
+  virtual void chain() {
+    using Eigen::Map;
+
+    matrix_d adj_w = Map<matrix_vi>(vari_ref_w_, M_, 1).adj();
+    matrix_d adj_v = Map<matrix_vi>(vari_ref_v_, M_, M_).adj();
+    Map<matrix_d> w(w_, M_, M_);
+    Map<matrix_d> v(v_, M_, M_);
+
+    // f(i,j) = (i != j ? 1 / (e_i - e_j) : 0).
```

Comment:
I think delete this comment. It's clear enough from the code. Just leaves me wondering what 'e' is (I guess it's the eigenvalues, but in the code this is 'w').

Reply:
Done.